pipelines/common/src/main/java/com/google/fhir/analytics/FetchUtil.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2020-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.fhir.analytics;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.rest.client.api.IClientInterceptor;
+import ca.uhn.fhir.rest.client.api.IGenericClient;
+import ca.uhn.fhir.rest.client.api.IHttpRequest;
+import ca.uhn.fhir.rest.client.interceptor.BasicAuthInterceptor;
+import ca.uhn.fhir.rest.client.interceptor.BearerTokenAuthInterceptor;
+import ca.uhn.fhir.rest.client.interceptor.LoggingInterceptor;
+import com.google.api.client.auth.oauth2.ClientCredentialsTokenRequest;
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.http.BasicAuthentication;
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import java.io.IOException;
+import java.time.Instant;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.Resource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A set of utility functions for making FHIR API calls. */
+public class FetchUtil {
+
+  private static final Logger log = LoggerFactory.getLogger(FetchUtil.class);
+
+  private final String fhirUrl;
+
+  private final String sourceUser;
+
+  private final String sourcePw;
+
+  private final String oAuthTokenEndpoint;
+
+  private final String oAuthClientId;
+
+  private final String oAuthClientSecret;
+
+  private final FhirContext fhirContext;
+
+  private final IClientInterceptor authInterceptor;
+
+  FetchUtil(
+      String sourceFhirUrl,
+      String sourceUser,
+      String sourcePw,
+      String oAuthTokenEndpoint,
```

Comment:
nit: Small refactoring - instead of the attributes `sourceUser`...`oAuthClientSecret` etc., can we pass the  `IClientInterceptor` by initializing it externally?

Reply:
I am not sure if that will make the code simpler. One benefit of the current approach is that it hides the notion of `IClientInterceptor` from the users of this class. If I want to change this to accept an `IClientInterceptor` instance then either the users of this need to get into the business of initializing interceptors (which I prefer not to) or I should provide helper methods that create one from these fields (which would be similar to what we have now).  WDYT?

Reply:
Makes sense as the client does not have initialize the interceptor.